<a href="https://colab.research.google.com/github/aluhanska/Data_Analyst-HW/blob/main/HW_16_1_%D0%9F%D1%80%D0%BE%D0%B3%D0%BD%D0%BE%D0%B7%D1%83%D0%B2%D0%B0%D0%BD%D0%BD%D1%8F_%D0%BE%D1%80%D0%B5%D0%BD%D0%B4%D0%BD%D0%BE%D1%97_%D0%BF%D0%BB%D0%B0%D1%82%D0%B8_%D0%B7%D0%B0_%D0%B6%D0%B8%D1%82%D0%BB%D0%BE.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Домашнє завдання: Прогнозування орендної плати за житло

## Мета завдання
Застосувати знання з лекції для побудови моделі лінійної регресії, що прогнозує орендну плату за житло в Індії. Ви пройдете весь цикл вирішення задачі машинного навчання: від дослідницького аналізу до оцінки якості моделі.

## Опис датасету
**House Rent Prediction Dataset** містить інформацію про 4700+ оголошень про оренду житла в Індії з такими параметрами:
- **BHK**: Кількість спалень, залів, кухонь
- **Rent**: Орендна плата (цільова змінна)
- **Size**: Площа в квадратних футах
- **Floor**: Поверх та загальна кількість поверхів
- **Area Type**: Тип розрахунку площі
- **Area Locality**: Район
- **City**: Місто
- **Furnishing Status**: Стан меблювання
- **Tenant Preferred**: Тип орендаря
- **Bathroom**: Кількість ванних кімнат
- **Point of Contact**: Контактна особа

---

## Завдання 1: Завантаження та перший огляд даних (1 бал)

**Що потрібно зробити:**
1. Завантажте дані з файлу `House_Rent_Dataset.csv`
2. Виведіть розмір датасету
3. Покажіть перші 5 рядків
4. Виведіть загальну інформацію про дані (включно з типами даних та кількістю значень)


In [1077]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [1078]:
import pandas as pd
import plotly.express as px
import numpy as np
import plotly.graph_objects as go
from plotly.subplots import make_subplots

In [1079]:
import numpy as np

In [1080]:
import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots

In [1081]:
# 1
df = pd.read_csv('/content/drive/MyDrive/Data Analytics/Аналітик даних в IT/House_Rent_Dataset.csv', sep=None, engine='python')

In [1082]:
# 2
df.shape

(4746, 12)

In [1083]:
# 3
df.head()

,Posted On,BHK,Rent,Size,Floor,Area Type,Area Locality,City,Furnishing Status,Tenant Preferred,Bathroom,Point of Contact
0,2022-05-18,2,10000,1100,Ground out of 2,Super Area,Bandel,Kolkata,Unfurnished,Bachelors/Family,2,Contact Owner
1,2022-05-13,2,20000,800,1 out of 3,Super Area,"Phool Bagan, Kankurgachi",Kolkata,Semi-Furnished,Bachelors/Family,1,Contact Owner
2,2022-05-16,2,17000,1000,1 out of 3,Super Area,Salt Lake City Sector 2,Kolkata,Semi-Furnished,Bachelors/Family,1,Contact Owner
3,2022-07-04,2,10000,800,1 out of 2,Super Area,Dumdum Park,Kolkata,Unfurnished,Bachelors/Family,1,Contact Owner
4,2022-05-09,2,7500,850,1 out of 2,Carpet Area,South Dum Dum,Kolkata,Unfurnished,Bachelors,1,Contact Owner


In [1084]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4746 entries, 0 to 4745
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   Posted On          4746 non-null   object
 1   BHK                4746 non-null   int64 
 2   Rent               4746 non-null   int64 
 3   Size               4746 non-null   int64 
 4   Floor              4746 non-null   object
 5   Area Type          4746 non-null   object
 6   Area Locality      4746 non-null   object
 7   City               4746 non-null   object
 8   Furnishing Status  4746 non-null   object
 9   Tenant Preferred   4746 non-null   object
 10  Bathroom           4746 non-null   int64 
 11  Point of Contact   4746 non-null   object
dtypes: int64(4), object(8)
memory usage: 445.1+ KB


## Завдання 2: Дослідницький аналіз даних (EDA) (5 балів)

**Що потрібно зробити:**
1. **Аналіз пропущених значень.** Перевірте наявність і відсоток пропущених значень у кожній колонці
2. **Базова статистика.** Обчисліть базову статистику (середнє, квартилі, стандартне відхилення) для числових змінних.
3. **Аналіз цільової змінної.** Побудуйте гістограму розподілу цільової змінної (Rent)
4. **Робота з викидами.** Знайдіть та видаліть викиди в цільовій змінній (якщо є). Визначити викиди можна будь-яким зрозумілим для вас способом, як варіант - таким, що використовується в побудові box-plot (https://en.wikipedia.org/wiki/Box_plot#Example_with_outliers).
5. **Аналіз категоріальних змінних.** Виведіть кількість унікальних значень для кожної з категоріальних колонок.


In [1085]:
# 1
missing_data = df.isnull().sum()
missing_data

,0
Posted On,0
BHK,0
Rent,0
Size,0
Floor,0
Area Type,0
Area Locality,0
City,0
Furnishing Status,0
Tenant Preferred,0


Відповідно, відсоток пропущених значень у кожній колонці дорівнює 0.

In [1086]:
# 2
stats = df[['BHK', 'Rent', 'Size', 'Bathroom']].describe().round(2)
stats

,BHK,Rent,Size,Bathroom
count,4746.00,4746.00,4746.00,4746.00
mean,2.08,34993.45,967.49,1.97
std,0.83,78106.41,634.20,0.88
min,1.00,1200.00,10.00,1.00
25%,2.00,10000.00,550.00,1.00
50%,2.00,16000.00,850.00,2.00
75%,3.00,33000.00,1200.00,2.00
max,6.00,3500000.00,8000.00,10.00


In [1087]:
# 3
fig = px.histogram(
    df,
    x='Rent',
    nbins=100,
    title='Розподіл цільової змінної (Rent)'
)
fig.update_layout(
    xaxis_title='Кількість Rent',
    yaxis_title='Кількість оголошень',
    height=400
)

Маємо правоскошений розподіл. Аби побачити його точніше давайте прологарифмуємо цільову змінну (Rent).

In [1088]:
df['log_Rent'] = np.log1p(df.Rent)

fig = px.histogram(
    df,
    x='log_Rent',
    nbins=100,
    title='Розподіл цільової змінної (log Rent)'
)
fig.update_layout(
    xaxis_title='log Rent',
    yaxis_title='Кількість оголошень',
    height=400
)

In [1089]:
df = df.drop('log_Rent', axis=1)

In [1090]:
# 4 Знаходимо межі викидів (IQR)

Q1 = df.Rent.quantile(0.25)
Q3 = df.Rent.quantile(0.75)
IQR = Q3 - Q1

lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

lower_bound, upper_bound

(np.float64(-24500.0), np.float64(67500.0))

In [1091]:
# 4 Знаходимо кількість викидів

outliers = df[(df.Rent < lower_bound) | (df.Rent > upper_bound)]
len(outliers)

520

In [1092]:
# 4 Видаляємо викиди

df_clean = df[(df.Rent >= lower_bound) & (df.Rent <= upper_bound)].copy()

In [1093]:
# 4 Перевіряємо результат

print("Було:", df.shape)
print("Стало:", df_clean.shape)

Було: (4746, 12)
Стало: (4226, 12)


In [1094]:
# 4 Додатково перевіряємо через box-plot (до)

fig = px.box(
    y=df.Rent,
    title='До видалення викидів',
    labels={'y': 'Rent'}
)
fig.show()

In [1095]:
# 4 Додатково перевіряємо через box-plot (після)

fig = px.box(
    y=df_clean.Rent,
    title='Після видалення викидів',
    labels={'y': 'Rent'}
)
fig.show()

In [1096]:
# 5
cat_cols = df_clean.select_dtypes('object').columns
df_clean[cat_cols].nunique()

,0
Posted On,80
Floor,340
Area Type,3
Area Locality,1997
City,6
Furnishing Status,3
Tenant Preferred,3
Point of Contact,3



## Завдання 3: Аналіз кореляцій та взаємозв'язків (3 бали)

**Що потрібно зробити:**
1. Обчисліть матрицю кореляцій для числових змінних
2. Візуалізуйте кореляційну матрицю за допомогою heatmap
3. Побудуйте scatter plot між Size та Rent
4. Проаналізуйте взаємозв'язок між BHK та Rent за допомогою boxplot (який розподіл плати для різних значень BHK)


In [1097]:
# 1
correlation_matrix = df_clean.select_dtypes('number').corr().round(2)
correlation_matrix

,BHK,Rent,Size,Bathroom
BHK,1.00,0.40,0.70,0.75
Rent,0.40,1.00,0.39,0.51
Size,0.70,0.39,1.00,0.68
Bathroom,0.75,0.51,0.68,1.00


In [1098]:
# 2

fig = px.imshow(
    correlation_matrix,
    text_auto='.2f',
    color_continuous_scale='RdBu_r',
    title='Кореляційна матриця для числових змінних',
    labels=dict(color="Кореляція")
)
fig.update_layout(height=500)

In [1099]:
# 3
fig = px.scatter(
    x=df_clean.Size,
    y=df_clean.Rent,
    title='Залежність Rent від Size',
    labels={'x': 'Size', 'y': 'Rent'},
    opacity=0.6
)
fig.show()

In [1100]:
# 4
fig = px.box(
    x=df_clean.BHK,
    y=df_clean.Rent,
    title='Розподіл орендної плати (Rent) залежно від BHK',
    labels={'x': 'BHK', 'y': 'Rent'}
)
fig.show()

## Завдання 4: Feature Engineering та підготовка даних (4 бали)

**Що потрібно зробити:**
1. Закодуйте категоріальні змінні за допомогою One-Hot Encoding. Пригадайте, що в лекції ми говорили щодо кодування кат. змінних з великої кількістю різних значень і як працювати з такими випадками. Ви можете закодувати не всі кат. змінні, а лише ті, що вважаєте за потрібні (скажімо ті, що мають відносно небагато різних значень).
2. **Опціонально (по 0.5 бала за кожну доцільну ознаку):** Додайте нові ознаки, обчислені на основі наявних даних, які б на ваш погляд були корисними для моделі
3. Виберіть ознаки для побудови моделі (виключіть непотрібні колонки). Виключити можна, наприклад, ті колонки, які мають категоріальний тип і забагато (більше 20) різних значень. Треба виключити хоча б 1 колонку.
4. Розділіть дані на ознаки (X) та цільову змінну (y)
5. Застосуйте стандартизацію до числових ознак


In [1101]:
# 1 Спочатку кодуємо категоріальні змінні, що мають відносно небагато різних значень (у нашому випадку <= 6)

type_dummies = pd.get_dummies(df_clean['Area Type'], prefix='type')
city_dummies = pd.get_dummies(df_clean['City'], prefix='city')
status_dummies = pd.get_dummies(df_clean['Furnishing Status'], prefix='status')
tenant_dummies = pd.get_dummies(df_clean['Tenant Preferred'], prefix='tenant')
point_of_dummies = pd.get_dummies(df_clean['Point of Contact'], prefix='point_of')

display(df_clean['Area Type'].head())                                 # перевіряємо вихідні дані (одні з)
display(type_dummies.head())                                          # перевіряємо правильність кодування (bool type)
display(type_dummies.head().astype(int))                              # перевіряємо правильність кодування (int type)

,Area Type
0,Super Area
1,Super Area
2,Super Area
3,Super Area
4,Carpet Area


,type_Built Area,type_Carpet Area,type_Super Area
0,False,False,True
1,False,False,True
2,False,False,True
3,False,False,True
4,False,True,False


,type_Built Area,type_Carpet Area,type_Super Area
0,0,0,1
1,0,0,1
2,0,0,1
3,0,0,1
4,0,1,0


In [1102]:
df_clean = pd.concat([df_clean, type_dummies, city_dummies, status_dummies, tenant_dummies, point_of_dummies], axis=1)

In [1103]:
# 2 Далі додаємо нові ознаки, обчислені на основі наявних даних:

# ознака № 1 - загальна кількість поверхів
# (орендар звертає увагу на поверховість будинку під час вибору квартири під оренду в першу чергу з огляду на наявність ліфта у високоповерховому будинку та його відсутність у низькоповерховому)

df_clean[['Floor_level', 'Total_floors']] = df_clean.Floor.str.split(' out of ', expand=True)
df_clean[['Floor_level', 'Total_floors']].head()

,Floor_level,Total_floors
0,Ground,2
1,1,3
2,1,3
3,1,2
4,1,2


In [1104]:
# ознака № 2 - поверх
# (поверховість квартири також має значення, наприклад, в середньому на високих поверхах кращий вигляд, а до 5 поверху можна відносно легко піднятися пішки)

df_clean['Floor_level'] = df_clean['Floor_level'].replace({
    'Ground': 0,
    'Upper Basement': -1,
    'Lower Basement': -2
})
df_clean['Floor_level'].head()

,Floor_level
0,0
1,1
2,1
3,1
4,1


In [1105]:
cat_cols = df_clean.select_dtypes('object').columns
df_clean[cat_cols].nunique()

,0
Posted On,80
Floor,340
Area Type,3
Area Locality,1997
City,6
Furnishing Status,3
Tenant Preferred,3
Point of Contact,3
Floor_level,32
Total_floors,47


In [1106]:
df_clean[['Total_floors', 'Floor_level']].isnull().sum()

,0
Total_floors,4
Floor_level,0


In [1107]:
df_clean[df_clean.isnull().any(axis=1)]

,Posted On,BHK,Rent,Size,Floor,Area Type,Area Locality,City,Furnishing Status,Tenant Preferred,...,status_Semi-Furnished,status_Unfurnished,tenant_Bachelors,tenant_Bachelors/Family,tenant_Family,point_of_Contact Agent,point_of_Contact Builder,point_of_Contact Owner,Floor_level,Total_floors
2553,2022-06-18,2,20000,400,3,Super Area,"Kasturba Niketan, Lajpat Nagar 2",Delhi,Unfurnished,Bachelors/Family,...,False,True,False,True,False,False,False,True,3,None
2883,2022-05-23,1,18000,450,Ground,Carpet Area,"DDA Flat AD Block, Shalimar Bagh AD Block",Delhi,Furnished,Bachelors/Family,...,False,False,False,True,False,False,False,True,0,None
4490,2022-06-12,3,15000,900,1,Super Area,"Malakpet, NH 9",Hyderabad,Semi-Furnished,Bachelors/Family,...,True,False,False,True,False,False,False,True,1,None
4560,2022-05-31,3,15000,1270,1,Carpet Area,Tarnaka,Hyderabad,Furnished,Family,...,False,False,False,False,True,False,False,True,1,None


In [1108]:
df_clean['Floor_level'] = pd.to_numeric(df_clean['Floor_level'], errors='coerce')
df_clean['Total_floors'] = pd.to_numeric(df_clean['Total_floors'], errors='coerce')

In [1109]:
df_clean['Total_floors'].describe().round(2)

,Total_floors
count,4222.00
mean,5.42
std,6.26
min,1.00
25%,2.00
50%,4.00
75%,5.00
max,68.00


In [1110]:
df_clean['Total_floors'] = df_clean['Total_floors'].fillna(df_clean['Total_floors'].median())

In [1111]:
df_clean['Total_floors'].isnull().sum()

np.int64(0)

In [1112]:
df_clean['Total_floors'].describe().round(2)

,Total_floors
count,4226.00
mean,5.42
std,6.26
min,1.00
25%,2.00
50%,4.00
75%,5.00
max,68.00


In [1113]:
# ознака 3 - нормалізована позиція поверху
# (необхідно для подальшого кодування даної категоріальної змінної)

def floor_cat(row):
  ratio = row['Floor_level'] / row['Total_floors']
  if ratio <= 0.33:
    return "low"
  elif ratio <= 0.66:
    return "middle"
  else:
    return "hight"

df_clean['floor_category'] = df_clean.apply(floor_cat, axis=1)
df_clean['floor_category']

,floor_category
0,low
1,middle
2,middle
3,middle
4,middle
...,...
4741,middle
4742,low
4743,middle
4744,hight


In [1114]:
floor_dummies = pd.get_dummies(df_clean['floor_category'], prefix='floor')
df_clean = pd.concat([df_clean, floor_dummies], axis=1)

In [1115]:
# обробляємо дату публікації оголошення з метою виявлення ознак, які можуть бути корисними для моделі
# та обираємо наступні ознаки, які можуть описувати сезонність та тижневі коливання активності користувачів сайту оголошень:
# - № 4 - квартал;
# - № 5 - місяць;
# - № 5 - день тижня;
# - № 6 - чи вихідний

df_clean['Posted On'] = pd.to_datetime(df_clean['Posted On'])

df_clean['year'] = df_clean['Posted On'].dt.year
df_clean['quarter'] = df_clean['Posted On'].dt.quarter
df_clean['month'] = df_clean['Posted On'].dt.month
df_clean['day'] = df_clean['Posted On'].dt.day
df_clean['week_of_year'] = df_clean['Posted On'].dt.isocalendar().week
df_clean['weekday'] = df_clean['Posted On'].dt.weekday                               # 0 - понеділок, 6 - неділя
df_clean['is_weekend'] = df_clean['Posted On'].dt.weekday.isin([5, 6]).astype(int)

df_clean[['year', 'quarter', 'month', 'day', 'week_of_year', 'weekday', 'is_weekend']].nunique()

,0
year,1
quarter,2
month,4
day,31
week_of_year,13
weekday,7
is_weekend,2


In [1116]:
# 3
df_clean = df_clean.drop(['Posted On', 'Floor', 'Area Type', 'Area Locality', 'City', 'Furnishing Status', 'Tenant Preferred', 'Point of Contact', 'Floor_level', 'Total_floors', 'floor_category', 'year', 'day', 'week_of_year'], axis=1)
df_clean.head()

,BHK,Rent,Size,Bathroom,type_Built Area,type_Carpet Area,type_Super Area,city_Bangalore,city_Chennai,city_Delhi,...,point_of_Contact Agent,point_of_Contact Builder,point_of_Contact Owner,floor_hight,floor_low,floor_middle,quarter,month,weekday,is_weekend
0,2,10000,1100,2,False,False,True,False,False,False,...,False,False,True,False,True,False,2,5,2,0
1,2,20000,800,1,False,False,True,False,False,False,...,False,False,True,False,False,True,2,5,4,0
2,2,17000,1000,1,False,False,True,False,False,False,...,False,False,True,False,False,True,2,5,0,0
3,2,10000,800,1,False,False,True,False,False,False,...,False,False,True,False,False,True,3,7,0,0
4,2,7500,850,1,False,True,False,False,False,False,...,False,False,True,False,False,True,2,5,0,0


In [1117]:
# 4
X = df_clean.drop(columns=['Rent'])
y = df_clean['Rent']
print(f"Розмір X (ознак): {X.shape}")
print(f"Розмір y (цільових змінних): {y.shape}")

Розмір X (ознак): (4226, 28)
Розмір y (цільових змінних): (4226,)


In [1118]:
from sklearn.preprocessing import StandardScaler

In [1119]:
# 5
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)
X_scaled_df = pd.DataFrame(X_scaled, columns=X.columns, index=X.index)
X_scaled_df.head()

,BHK,Size,Bathroom,type_Built Area,type_Carpet Area,type_Super Area,city_Bangalore,city_Chennai,city_Delhi,city_Hyderabad,...,point_of_Contact Agent,point_of_Contact Builder,point_of_Contact Owner,floor_hight,floor_low,floor_middle,quarter,month,weekday,is_weekend
0,0.052966,0.469859,0.272578,-0.02176,-0.878583,0.879429,-0.502144,-0.501774,-0.387614,-0.499926,...,-0.577532,-0.015385,0.577897,-0.672483,1.419248,-0.744862,-0.484717,-0.861071,-0.509557,-0.561499
1,0.052966,-0.147778,-1.133910,-0.02176,-0.878583,0.879429,-0.502144,-0.501774,-0.387614,-0.499926,...,-0.577532,-0.015385,0.577897,-0.672483,-0.704598,1.342530,-0.484717,-0.861071,0.578562,-0.561499
2,0.052966,0.263980,-1.133910,-0.02176,-0.878583,0.879429,-0.502144,-0.501774,-0.387614,-0.499926,...,-0.577532,-0.015385,0.577897,-0.672483,-0.704598,1.342530,-0.484717,-0.861071,-1.597675,-0.561499
3,0.052966,-0.147778,-1.133910,-0.02176,-0.878583,0.879429,-0.502144,-0.501774,-0.387614,-0.499926,...,-0.577532,-0.015385,0.577897,-0.672483,-0.704598,1.342530,2.063061,1.546396,-1.597675,-0.561499
4,0.052966,-0.044839,-1.133910,-0.02176,1.138197,-1.137102,-0.502144,-0.501774,-0.387614,-0.499926,...,-0.577532,-0.015385,0.577897,-0.672483,-0.704598,1.342530,-0.484717,-0.861071,-1.597675,-0.561499


## Завдання 5: Розділення даних та навчання моделі (3 бали)

**Що потрібно зробити:**
1. Розділіть дані на навчальну (80%) та тестову (20%) вибірки.
2. Створіть модель лінійної регресії.
3. Навчіть модель на навчальних даних.
4. Виведіть усі коефіцієнти моделі (ваги) та напишіть, які 2 ознаки найбільше впливають на прогноз.
5. Зробіть прогнози на тренувальній та тестовій вибірках.

In [1120]:
from sklearn.model_selection import train_test_split

In [1121]:
# 1
X_train, X_test, y_train, y_test = train_test_split(
    X_scaled_df, y,
    test_size=0.2,
    random_state=42
)

In [1122]:
from sklearn.linear_model import LinearRegression

In [1123]:
# 2
model = LinearRegression()

In [1124]:
# 3
model.fit(X_train, y_train)

LinearRegression()

In [1125]:
# 4
importance = pd.DataFrame({
    'feature': model.feature_names_in_,
    'coef': model.coef_.round(2)
})

importance['abs_coef'] = np.abs(importance.coef)
importance_sorted = importance.sort_values(by='abs_coef', ascending=False)
importance_sorted

,feature,coef,abs_coef
11,city_Mumbai,5844.71,5844.71
1,Size,3862.65,3862.65
0,BHK,2610.34,2610.34
18,point_of_Contact Agent,1862.73,1862.73
20,point_of_Contact Owner,-1861.45,1861.45
2,Bathroom,1817.01,1817.01
9,city_Hyderabad,-1751.78,1751.78
10,city_Kolkata,-1688.47,1688.47
7,city_Chennai,-1259.12,1259.12
12,status_Furnished,1069.47,1069.47


2 ознаки, які найбільше впливають на прогноз: `city_Mumbai` та `Size	`.


In [1126]:
# 5

y_train_pred = model.predict(X_train)
y_test_pred = model.predict(X_test)

comparison_train = pd.DataFrame({
    'Реальна Rent': y_train.values,
    'Прогнозована Rent': y_train_pred.round(2),
    'Помилка': (y_train.values - y_train_pred).round(2)
})
print("ПРИКЛАД ПРОГНОЗУ НА НАВЧАЛЬНІЙ ВИБІРЦІ:")
print(comparison_train[:5])

comparison_test = pd.DataFrame({
    'Реальна Rent': y_test.values,
    'Прогнозована Rent': y_test_pred.round(2),
    'Помилка': (y_test.values - y_test_pred).round(2)
})
print("\nПРИКЛАД ПРОГНОЗУ НА ТЕСТОВІЙ ВИБІРЦІ:")
print(comparison_test[:5])

ПРИКЛАД ПРОГНОЗУ НА НАВЧАЛЬНІЙ ВИБІРЦІ:
   Реальна Rent  Прогнозована Rent  Помилка
0         24000           21151.19  2848.81
1         28000           27454.35   545.65
2         13000           12490.70   509.30
3         18500           15983.56  2516.44
4         15000           17039.16 -2039.16

ПРИКЛАД ПРОГНОЗУ НА ТЕСТОВІЙ ВИБІРЦІ:
   Реальна Rent  Прогнозована Rent  Помилка
0         22000           29357.35 -7357.35
1          5000            3437.16  1562.84
2         37000           40067.13 -3067.13
3          8000            2810.70  5189.30
4         15000           15103.57  -103.57


## Завдання 6: Оцінка якості моделі (2 бали)

**Що потрібно зробити:**
1. Обчисліть MAE, RMSE та R² для навчальної та тестової вибірок
2. Порівняйте метрики та зробіть висновок про якість моделі
3. Проаналізуйте і дайте висновок, чи є ознаки перенавчання або недонавчання (**Нагадування**: перенавчання - коли модель дуже добре працює на тренувальних даних, але погано на тестових; недонавчання - коли модель погано працює навіть на тренувальних даних)
4. Побудуйте графік розсіювання "реальні vs прогнозовані значення" та зробіть висновок про якість моделі


In [1127]:
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

In [1128]:
# 1

mae = mean_absolute_error(y_train, y_train_pred)
rmse = np.sqrt(mean_squared_error(y_train, y_train_pred))
r2 = r2_score(y_train, y_train_pred)

print("МЕТРИКИ ЯКОСТІ МОДЕЛІ ДЛЯ НАВЧАЛЬНОЇ ВИБІРКИ:\n")
print(f"MAE: {mae:.2f} орендних плат")
print(f"RMSE: {rmse:.2f} орендних плат")
print(f"R²: {r2:.3f}")

mae = mean_absolute_error(y_test, y_test_pred)
rmse = np.sqrt(mean_squared_error(y_test, y_test_pred))
r2 = r2_score(y_test, y_test_pred)

print("\nМЕТРИКИ ЯКОСТІ МОДЕЛІ ДЛЯ ТЕСТОВОЇ ВИБІРКИ:")
print(f"MAE: {mae:.2f} орендних плат")
print(f"RMSE: {rmse:.2f} орендних плат")
print(f"R²: {r2:.3f}")

МЕТРИКИ ЯКОСТІ МОДЕЛІ ДЛЯ НАВЧАЛЬНОЇ ВИБІРКИ:

MAE: 5539.40 орендних плат
RMSE: 7681.80 орендних плат
R²: 0.691

МЕТРИКИ ЯКОСТІ МОДЕЛІ ДЛЯ ТЕСТОВОЇ ВИБІРКИ:
MAE: 5507.40 орендних плат
RMSE: 7694.85 орендних плат
R²: 0.691


In [1129]:
mean_train = np.mean(y_train).round(2)
mean_test = np.mean(y_test).round(2)

print(f"Середнє значення для НАВЧАЛЬНОЇ вибірки: {mean_train}")
print(f"Середнє значення для ТЕСТОВОЇ вибірки: {mean_test}")

Середнє значення для НАВЧАЛЬНОЇ вибірки: 19349.55
Середнє значення для ТЕСТОВОЇ вибірки: 19032.92


2. З огляду на те, що для обох вибірок середнє значення більше за **MAE**, **RMSE** > **MAE**, **R²** (0.691) тяжіє до 1, то якість моделі доволі непогана.

3. Так як **R²** для навчальної вибірки = **R²** для тестової вибірки, то ознак перенавчання або недовчання не виявлено.

In [1130]:
# 4 Для навчальної вибірки:

fig = px.scatter(
    x=y_train,
    y=y_train_pred,
    title='Реальна vs Прогнозована Rent (навчальна вибірка)',
    labels={'x': 'Реальна Rent', 'y': 'Прогнозована Rent'},
    opacity = 0.6
)

max_val = max(y_train.max(), y_train_pred.max())
fig.add_trace(
    go.Scatter(
        x=[0, max_val],
        y=[0, max_val],
        mode='lines',
        name='Ідеальний прогноз',
        line=dict(color='red', dash='dash')
    )
)

In [1131]:
# Для тестової вибірки:

fig = px.scatter(
    x=y_test,
    y=y_test_pred,
    title='Реальна vs Прогнозована Rent (тестова вибірка)',
    labels={'x': 'Реальна Rent', 'y': 'Прогнозована Rent'},
    opacity = 0.6
)

max_val = max(y_test.max(), y_test_pred.max())
fig.add_trace(
    go.Scatter(
        x=[0, max_val],
        y=[0, max_val],
        mode='lines',
        name='Ідеальний прогноз',
        line=dict(color='red', dash='dash')
    )
)

В цільому модель гарна, але схильна переоцінювати орендну плату (`Rent`) за низьких значень та недооцінювати за високих.

## Завдання 7: Аналіз помилок (4 бали)

**Що потрібно зробити:**
1. Обчисліть помилки (residuals = реальні - прогнозовані значення)
2. Побудуйте гістограму розподілу помилок
3. Створіть scatter plot помилок відносно величини прогнозованих значень. Чи росте помилка з ростом прогнозованого значення?
4. Знайдіть 5 прогнозів з найбільшими помилками
5. Проаналізуйте, на яких типах житла модель помиляється найбільше. Типи можна розрізняти за кількістю кімнат чи містом, наприклад.
6. Подумайте і напишіть, які наступні кроки ви б зробили, аби поліпшити якість моделі. Опціонально можна їх зробити і ми перевіримо :)

In [1132]:
# 1
residuals_train = y_train - y_train_pred
residuals_test = y_test - y_test_pred

In [1133]:
# 2 Для навчальної вибірки:

fig = px.histogram(
    x=residuals_train,
    nbins=50,
    title = 'Розподіл помилок прогнозування (навчальна вибірка)',
    color_discrete_sequence=['#e74c3c']
)
fig.add_vline(
    x=0,
    line_dash='dash',
    line_color='black',
    annotation_text='Ідеальний прогноз'
)
fig.update_layout(
    xaxis_title='Помилка',
    yaxis_title='Кількість',
    height=400
)

In [1134]:
# 2 Для тестової вибірки:

fig = px.histogram(
    x=residuals_test,
    nbins=50,
    title = 'Розподіл помилок прогнозування (тестова вибірка)',
    color_discrete_sequence=['#e74c3c']
)
fig.add_vline(
    x=0,
    line_dash='dash',
    line_color='black',
    annotation_text='Ідеальний прогноз'
)
fig.update_layout(
    xaxis_title='Помилка',
    yaxis_title='Кількість',
    height=400
)

In [1135]:
# 3 Для навчальної вибірки:
fig = px.scatter(
    x=residuals_train,
    y=y_train,
    title='Залежність прогнозованої Rent від помилки (навчальна вибірка)',
    labels={'x': 'Помилка', 'y': 'Прогнозована Rent'},
    opacity = 0.6
)
fig.show()

In [1136]:
# 3 Для тестової вибірки:
fig = px.scatter(
    x=residuals_test,
    y=y_test,
    title='Залежність прогнозованої Rent від помилки (тестова вибірка)',
    labels={'x': 'Помилка', 'y': 'Прогнозована Rent'},
    opacity = 0.6
)
fig.show()

Як на видно на графіках, помилка росте із ростом прогнозованого значення `Rent`.

In [1137]:
# 4

comparison_train['Помилка'] = comparison_train['Помилка'].abs()
comparison_test['Помилка'] = comparison_test['Помилка'].abs()

comparison_train = comparison_train.sort_values(by='Помилка', ascending=False)
comparison_test = comparison_test.sort_values(by='Помилка', ascending=False)

print("ТОП-5 ПРОГНОЗІВ З НАЙБІЛЬШИМИ ПОМИЛКАМИ НА НАВЧАЛЬНІЙ ВИБІРЦІ:")
print(comparison_train[:5])

print("\nТОП-5 ПРОГНОЗІВ З НАЙБІЛЬШИМИ ПОМИЛКАМИ НА ТЕСТОВІЙ ВИБІРЦІ:")
print(comparison_test[:5])

ТОП-5 ПРОГНОЗІВ З НАЙБІЛЬШИМИ ПОМИЛКАМИ НА НАВЧАЛЬНІЙ ВИБІРЦІ:
      Реальна Rent  Прогнозована Rent   Помилка
311          65000           27626.24  37373.76
128          52000           15093.38  36906.62
2846         65000           28623.92  36376.08
272          55000           20453.03  34546.97
2961         60000           26011.74  33988.26

ТОП-5 ПРОГНОЗІВ З НАЙБІЛЬШИМИ ПОМИЛКАМИ НА ТЕСТОВІЙ ВИБІРЦІ:
     Реальна Rent  Прогнозована Rent   Помилка
642         65000           32618.22  32381.78
193          8000           39615.54  31615.54
122         30000           59942.98  29942.98
210          8000           37344.03  29344.03
337         65000           35925.79  29074.21


In [1138]:
# 5 Для навчальної вибірки:

comparison_train['BHK'] = X_train['BHK']

bhk_analysis_train = comparison_train.groupby('BHK')['Помилка'].median().sort_values(ascending=False)
bhk_analysis_train

,Помилка
BHK,
2.733610,4642.720
-1.287356,4213.610
0.052966,3844.720
1.393288,3656.390
4.073931,2820.845
5.414253,1003.260


In [1139]:
# Для тестової вибірки:

comparison_test['BHK'] = X_test['BHK']

bhk_analysis_test = comparison_test.groupby('BHK')['Помилка'].median().sort_values(ascending=False)
bhk_analysis_test

,Помилка
BHK,
2.733610,7970.95
-1.287356,3978.52
0.052966,3915.92
1.393288,2766.01


Модель помиляється найбільше за кількістю кімнат від 2 до 3.

6. Аби ще поліпшити якість моделі можливо я все ж таки врахувала значення `day` та `week_of_year`, але це не точно :) так як вони мені здаються геть не логічними.